text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,464 @@
+While in reality, this can either cause unexpected errors or non-working features.
+
+We want to come up with a consistent model that users can rely on.
+
+## Simplify CX Protocol Versioning
+
+Recently, we have had a few customer impacting issues that relate to our cx-protocol.
+
+- [aws/aws-cdk: A newer version of the CDK CLI (>= 1.21.0) is necessary to interact with this app](https://github.com/aws/aws-cdk/issues/5986)
+- [aws/aws-cdk: CDK CLI can only be used with apps created by CDK >= 1.10.0](https://github.com/aws/aws-cdk/issues/4294)
+
+A major contributing factor to those issues, is that the ergonomics of introducing breaking changes (cx-protocol version bump), are not very good.
+Something about the process feels convoluted and we seem to be getting it wrong almost every time we attempt it.
+
+We should consider a re-think of the entire process, as well as the assumptions that lead to it.
+
+## Validate Compatibility
+
+We want to make sure we have sufficient testing in place that validate we don't introduce any accidental CLI compatibility breakage.
+
+**Currently, the only way for us to detect these types of problems is by noticing them in the code review process.**
+
+# Design Summary
+
+Assuming we adopt the new compatibility model, all we have to do is come up with a validation mechanism for the following breakage scenarios:
+
+## Breaking changes in CX Protocol
+
+Probably the most prominent scenario for causing breakage is pushing an incompatible change to the cx-protocol itself.
+This can, for example, be the removal of a property.
+
+> See concrete [example](#rename-target-property-in-containerImageAssetMetadataEntry).
+
+The proposed solution for this case is to run API compatibility checks using `jsii-diff`.
+We would treat `cx-api` as a regular `jsii` module that needs to maintain compatibility to its consumers, which is the CLI.
+
+This will make sure that when the CLI pulls in new versions of `cx-api` (on upgrades), it will properly consume older cloud assemblies.
+
+## Breaking changes in CLI
+
+Another way to break CLI upgrades, is to simply change the CLI in a breaking way. This can, for example, be a change to the docker build command.
+
+> See concrete [example](#remove---target-from-docker-build-command).
+
+The proposed solution for this case is to run standard regression tests. That is, run old CLI integration tests on the latest code (Framework and CLI).
+
+This will make sure that old CLI functionality is still working, assuming of course it was covered by our integration tests.
+
+# Detailed Design
+
+There are 2 mechanisms we need to implement:
+
+## API Compatibility Checks
+
+We need to make sure the objects that make up what we refer to as the `cx-protocol`, do not change in a breaking way.
+We should be able to leverage `jsii` to accomplish this. For example, this is an excerpt from the `.jsii` spec of the `cx-api` module:
```

Comment:
Worth explicitly noting that in order to leverage jsii compatibility checks for the Cloud Assembly interface, the module needs to use the cloud assembly as an "input".

That is, something like this:

```ts
export interface CloudAssembly { ... }

export class Dummy {
  public static dummy(assembly: CloudAssembly): void { }
  private constructor() { }
}
```

When `jsii-diff` is executed against this API, it will enforce that only non-breaking changes (i.e. addition of only optional attributes for example) will be introduced to the `CloudAssembly` interface.

Reply:
see my [comment](https://github.com/aws/aws-cdk-rfcs/pull/111#discussion_r380867816)